main/src/ca/uwaterloo/flix/language/phase/jvm/BackendObjType.scala

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2021 Jonathan Lindegaard Starup
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.jvm
+
+import ca.uwaterloo.flix.language.ast.Symbol
+import ca.uwaterloo.flix.language.phase.jvm.JvmName.{Delimiter, DevFlixRuntime, JavaLang, RootPackage}
+
+/**
+  * Represents all Flix types that are objects on the JVM (array is an exception).
+  */
+sealed trait BackendObjType {
+  /**
+    * Constructs a concatenated list of erased strings delimited with `JvmName.Delimiter`.
+    */
+  private def erasedListOfTypes(ts: List[BackendType]): String =
+    ts.map(e => e.toErased.toString).mkString(Delimiter)
+
+  val jvmName: JvmName = this match {
+    case BackendObjType.Unit => JvmName(DevFlixRuntime, "Unit")
+    case BackendObjType.BigInt => JvmName(List("java", "math"), "BigInteger")
+    case BackendObjType.String => JvmName(JavaLang, "String")
+    case BackendObjType.Channel(_) => JvmName(List("ca", "uwaterloo", "flix", "runtime", "interpreter"), "Channel")
+    case BackendObjType.Lazy(tpe) => JvmName(RootPackage, s"Lazy$Delimiter${tpe.toErased}")
+    case BackendObjType.Ref(tpe) => JvmName(RootPackage, s"Ref$Delimiter${tpe.toErased}")
+    case BackendObjType.Tuple(elms) => JvmName(RootPackage, s"Tuple${elms.length}$Delimiter${erasedListOfTypes(elms)}")
+    case BackendObjType.Enum(_, _) => JvmName(RootPackage, "MissingInBackendObjType") // TODO
+    case BackendObjType.Arrow(args, result) => JvmName(RootPackage, s"Fn${args.length}$Delimiter${erasedListOfTypes(args)}$Delimiter${result.toErased}")
+    case BackendObjType.RecordEmpty => JvmName(RootPackage, s"RecordEmpty$Delimiter")
+    case BackendObjType.RecordExtend(_, value, _) => JvmName(RootPackage, s"RecordExtend$Delimiter${value.toErased}")
+    case BackendObjType.SchemaEmpty => JvmName(RootPackage, "MissingInBackendObjType") // TODO
+    case BackendObjType.SchemaExtend(_, _, _) => JvmName(RootPackage, "MissingInBackendObjType") // TODO
+    case BackendObjType.Relation(_) => JvmName(RootPackage, "MissingInBackendObjType") // TODO
+    case BackendObjType.Lattice(_) => JvmName(RootPackage, "MissingInBackendObjType") // TODO
```

Comment:
Why not throw exceptions instead? I would prefer exceptions over silent breakage.

Reply:
Yeah me too, but it breaks since `SchemaEmpty` was an object and wanted its values. But I've made it a class temporarily